packages/valko-ui-components/src/components/Datepicker.vue

```diff
@@ -0,0 +1,307 @@
+import useStyle from '#valkoui/composables/useStyle.ts'
+import VkInput from './Input.vue'
+import VkButton from './Button.vue'
+import VkIcon from './Icon.vue'
+import VkPopover from './Popover.vue'
+import dayjs from 'dayjs'
+
+defineOptions({ name: 'VkDatepicker' })
+
+const props = withDefaults(defineProps<DatepickerProps>(), {
+  color: 'primary',
+  variant: 'filled',
+  size: 'md',
+  shape: 'soft'
+})
+
+const classes = useStyle<DatepickerProps, SlotStyles>(props, styles)
+
+// variable declarations
+const isOpen = ref(false)
+const dateFormat = 'DD-MM-YYYY'
+const currentDate = ref(dayjs().format(dateFormat))
+const currentMonth = ref(dayjs().month())
+const currentYear = ref(dayjs().year())
+const currentView: Ref<viewType> = ref('days')
+const viewDisplay = ref<string>('')
+const yearsList = ref<Array<number>>([])
+const filteredDaysList = ref<Array<number>>([])
+const leadingEmptyDays = ref<number>(0)
+const trailingEMptyDays = ref<number>(0)
+
+const toggleCalendar = () => {
+  isOpen.value = !isOpen.value
+}
+
+// Main functions to handle the date selection in the views
+const onDateSelect = (day: number | null) => {
+  if (day === null) return
+
+  const date = dayjs().year(currentYear.value).month(currentMonth.value).date(day)
+
+  currentDate.value = date.format(dateFormat)
+  currentMonth.value = date.month()
+  currentYear.value = date.year()
+
+  updateDisplay()
+  isOpen.value = false
+}
+
+const setMonth = (month: number) => {
+  currentMonth.value = month
+  currentView.value = 'days'
+}
+
+const setYear = (year: number) => {
+  currentYear.value = year
+  currentView.value = 'months'
+}
+
+// Handler for displayed text in the main action button
+const updateDisplay = () => {
+  if (currentView.value === 'days') viewDisplay.value = `${monthNames[currentMonth.value]} ${currentYear.value}`
+  if (currentView.value === 'months') viewDisplay.value = `${currentYear.value}`
+
+  if (currentView.value === 'years') {
+    const startYear = Math.floor(currentYear.value / 10) * 10
+    const endYear = startYear + 9
+    viewDisplay.value = `${startYear} - ${endYear}`
+  }
+}
+
+// Handler for view change
+const changeView = () => {
+  if (currentView.value === 'days') return currentView.value = 'months'
+  if (currentView.value === 'months') return currentView.value = 'years'
+  if (currentView.value === 'years') return currentView.value = 'days'
+}
+
+// Nagivation arrows actions
+const nagivateViewMap = (direction: number) => {
+  const actionsMap: Record<viewType, () => void> = {
+    days: () => {
+      currentMonth.value += direction
+
+      if (currentMonth.value < 0) {
+        currentMonth.value = 11
+        currentYear.value -= 1
+      }
+
+      if (currentMonth.value > 11) {
+        currentMonth.value = 0
+        currentYear.value += 1
+      }
+    },
+    months: () => currentYear.value += direction,
+    years: () => currentYear.value += direction * 10
+  }
+
+  actionsMap[currentView.value]()
+}
```

Comment:
a function inside a function inside a function, please find a way to refactor this, maybe we can create some helper functions or a composable to move all the formatting logic to other place